File: CCIP/CCIP/AnnounceTableViewController.m
//
//  AnnounceTableViewController.m
//  CCIP
//
//  Created by Sars on 8/10/16.
//  Copyright © 2016 CPRTeam. All rights reserved.
//

#import "AppDelegate.h"
#import "AnnounceTableViewController.h"
#import "AnnounceTableViewCell.h"
#import <SafariServices/SafariServices.h>
#import <AFNetworking/AFNetworking.h>
#import "WebServiceEndPoint.h"
#import "UIColor+addition.h"
#import "UITableView+FDTemplateLayoutCell.h"
#import "UIView+FDCollapsibleConstraints.h"
#import "UIView+addition.h"

@interface AnnounceTableViewController ()

@property (strong, nonatomic) FBShimmeringView *shimmeringLogoView;

@property (strong, nonatomic) UIRefreshControl *refreshControl;

@property (readwrite, nonatomic) BOOL loaded;

@property (readwrite, nonatomic) CGFloat controllerTopStart;

@end

@implementation AnnounceTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.announceTableView setSeparatorColor:[UIColor clearColor]];
    self.announceJsonArray = @[];
    if (self.refreshControl == nil) {
        self.refreshControl = [[UIRefreshControl alloc] init];
        [self.refreshControl addTarget:self
                                action:@selector(refresh)
                      forControlEvents:UIControlEventValueChanged];
        [self.announceTableView addSubview:self.refreshControl];
    }
    
    [self.navigationItem setTitle:NSLocalizedString(@"AnnouncementTitle", nil)];
    [self.navigationController.navigationBar setBackgroundImage:[UIImage new] forBarMetrics:UIBarMetricsDefault];
    [self.navigationController.navigationBar setShadowImage:[UIImage new]];
    [self.navigationController.navigationBar setBackgroundColor:[UIColor clearColor]];
    CGRect frame = CGRectMake(0, 0, self.view.frame.size.width, 239);
    UIView *headView = [UIView new];
    [headView setFrame:frame];
    [headView setGradientColor:[UIColor colorFromHtmlColor:@"#F9FEA5"]
                            To:[UIColor colorFromHtmlColor:@"#20E2D7"]
                    StartPoint:CGPointMake(-.4f, .5f)
                       ToPoint:CGPointMake(1, .5f)];
    [self.view addSubview:headView];
    [self.view sendSubviewToBack:headView];
    
    NSString *noAnnouncementText = NSLocalizedString(@"NoAnnouncementText", nil);
    NSMutableAttributedString *attributedNoAnnouncementText = [[NSMutableAttributedString alloc] initWithString:noAnnouncementText];
    float spacing = 5.0f;
    [attributedNoAnnouncementText addAttribute:NSKernAttributeName
                             value:@(spacing)
                             range:NSMakeRange(0, [noAnnouncementText length])];
    [self.lbNoAnnouncement setAttributedText:attributedNoAnnouncementText];
    
    SEND_GAI(@"AnnounceTableViewController");
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.controllerTopStart = self.navigationController.navigationBar.frame.size.height;
    [AppDelegate setDevLogo:self.shimmeringLogoView
                   WithLogo:ASSETS_IMAGE(@"AssetsUI", @"coscup-logo")];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self refresh];
}

- (void)refresh {
    self.loaded = NO;
    [self.refreshControl beginRefreshing];
    AFHTTPSessionManager *manager = [AFHTTPSessionManager manager];
    [manager GET:CC_ANNOUNCEMENT parameters:nil progress:nil success:^(NSURLSessionTask *task, id responseObject) {
        NSLog(@"JSON: %@", responseObject);
        if (responseObject != nil) {
            self.loaded = YES;
            self.announceJsonArray = responseObject;
            [self.announceTableView reloadData];
            [self.refreshControl endRefreshing];
        }
    } failure:^(NSURLSessionTask *operation, NSError *error) {
        NSLog(@"Error: %@", error);
        if ([error code] == -1009) {
            [self performSegueWithIdentifier:@"ShowInvalidNetworkMsg"
                                      sender:NSLocalizedString(@"Networking_Broken", nil)];
        } else {
            self.loaded = YES;
            self.announceJsonArray = @[];
            [self.announceTableView reloadData];
        }
        [self.refreshControl endRefreshing];
    }];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    UIViewController *destination = segue.destinationViewController;
    if ([destination isMemberOfClass:[InvalidNetworkMessageViewController class]]) {
        InvalidNetworkMessageViewController *inmvc = (InvalidNetworkMessageViewController *)destination;
        [inmvc setMessage:sender];
        [inmvc setDelegate:self];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

# pragma mark - UITableViewControllerDelegate
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSInteger count = [self.announceJsonArray count];
    if (self.loaded) {
        BOOL NoAnnouncement = count == 0;
        [self.ivNoAnnouncement setHidden:!NoAnnouncement];
        [self.lbNoAnnouncement setHidden:!NoAnnouncement];
    } else {
        [self.ivNoAnnouncement setHidden:YES];
        [self.lbNoAnnouncement setHidden:YES];
    }
    return count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    AnnounceTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"AnnounceCell" forIndexPath:indexPath];
    
    [self configureCell:cell atIndexPath:indexPath];
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return [tableView fd_heightForCellWithIdentifier:@"AnnounceCell" configuration:^(id cell) {
        [self configureCell:cell atIndexPath:indexPath];
    }];
}

- (void)configureCell:(AnnounceTableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath {
    [cell setFd_enforceFrameLayout:NO]; // Enable to use "-sizeThatFits:"
    [cell setSelectionStyle:UITableViewCellSelectionStyleNone];
    [cell setClipsToBounds:NO];
    [cell setBackgroundColor:[UIColor clearColor]];
    [cell.layer setZPosition:indexPath.row];
    UIView *vwContent = [cell performSelector:@selector(vwContent)];
    [vwContent.layer setCornerRadius:5.0f];
    [vwContent.layer setMasksToBounds:YES];
    
    UIView *vwShadowContent = [cell performSelector:@selector(vwShadowContent)];
    [vwShadowContent.layer setCornerRadius:5.0f];
    [vwShadowContent.layer setMasksToBounds:NO];
    [vwShadowContent.layer setShadowRadius:50.0f];
    [vwShadowContent.layer setShadowOffset:CGSizeMake(0, 50)];
    [vwShadowContent.layer setShadowColor:[[UIColor blackColor] CGColor]];
    [vwShadowContent.layer setShadowOpacity:0.1f];
    
    NSDictionary *announce = [self.announceJsonArray objectAtIndex:indexPath.row];
    NSString* language = [[[NSBundle mainBundle] preferredLocalizations] objectAtIndex:0];
    
    if ([language containsString:@"zh"]) {
        [cell.lbMessage setText:[announce objectForKey:@"msg_zh"]];
    } else {
        [cell.lbMessage setText:[announce objectForKey:@"msg_en"]];
    }
    NSString *uri = [announce objectForKey:@"uri"];
    BOOL hasURL = !(!uri || [uri isEqualToString:@""]);
    //    [cell setAccessoryType:hasURL ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone];
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm"];
    NSNumber *datetime = [announce objectForKey:@"datetime"];
    NSString *strDate = [dateFormatter stringFromDate:[NSDate dateWithTimeIntervalSince1970:[datetime doubleValue]]];
    
    [cell.lbMessageTime setText:strDate];
    
    if (hasURL) {
        [cell.lbURL setText:uri];
    } else {
        [cell.lbURL setText:@""];
    }
    [cell.vwDashedLine addDashedLine:[UIColor colorFromHtmlColor:@"#E9E9E9"]];
    
    [cell.vwURL setFd_collapsed:!hasURL];
}

#pragma mark - Table view delegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    NSDictionary *announce = [self.announceJsonArray objectAtIndex:indexPath.row];
    NSString *uri = [announce objectForKey:@"uri"];
    
    if (!uri || [uri isEqualToString:@""]) return;
    
    if ([SFSafariViewController class] != nil) {
        SFSafariViewController *safariViewController = [[SFSafariViewController alloc] initWithURL:[NSURL URLWithString:uri]];
        [[UIApplication getMostTopPresentedViewController] presentViewController:safariViewController animated:YES completion:nil];
    } else {
        if (![[UIApplication sharedApplication] openURL:[NSURL URLWithString:uri]]) {
            NSLog(@"%@%@",@"Failed to open url:", [[NSURL URLWithString:uri] description]);
        }
    }
    
    SEND_GAI_EVENT(@"AnnounceTableView", uri);
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: OPass/Tabs/AnnounceTab/AnnounceTableViewController.swift
//
//  AnnounceTableViewController.swift
//  OPass
//
//  Created by 腹黒い茶 on 2019/3/16.
//  2019 OPass.
//

import Foundation
import UIKit
import SwiftDate
import SwiftyJSON
import UIView_FDCollapsibleConstraints
import UITableView_FDTemplateLayoutCell

class AnnounceTableViewController: UIViewController, InvalidNetworkRetryDelegate, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet public var announceTableView: UITableView!
    @IBOutlet public var ivNoAnnouncement: UIImageView!
    @IBOutlet public var lbNoAnnouncement: UILabel!
    public var announceJsonArray: [AnnouncementInfo] = [AnnouncementInfo]()

    private var refreshControl: UIRefreshControl = UIRefreshControl.init()
    private var loaded: Bool = false
    @objc public var controllerTopStart: CGFloat = 0.0

    override func viewDidLoad() {
        super.viewDidLoad()
        self.announceTableView.separatorColor = UIColor.clear
        self.announceJsonArray = [AnnouncementInfo]()
        self.refreshControl.addTarget(self, action: #selector(refresh), for: .valueChanged)
        self.announceTableView.addSubview(self.refreshControl)

        self.navigationItem.title = NSLocalizedString("AnnouncementTitle", comment: "")
        self.navigationController?.navigationBar.setBackgroundImage(UIImage.init(), for: .default)
        self.navigationController?.navigationBar.shadowImage = UIImage.init()
        self.navigationController?.navigationBar.backgroundColor = UIColor.clear
        let frame = CGRect(x: 0, y: 0, width: self.view.frame.size.width, height: 239)
        let headView = UIView.init(frame: frame)
        headView.setGradientColor(
            from: Constants.appConfigColor.AnnouncementTitleLeftColor,
            to: Constants.appConfigColor.AnnouncementTitleRightColor,
            startPoint: CGPoint(x: -0.4, y: -0.5),
            toPoint: CGPoint(x: 1, y: 0.5)
        )
        self.view.addSubview(headView)
        self.view.sendSubviewToBack(headView)

        let noAnnouncementText = NSLocalizedString("NoAnnouncementText", comment: "")
        let attributedNoAnnouncementText = NSMutableAttributedString.init(string: noAnnouncementText)
        attributedNoAnnouncementText.addAttributes(
            [ NSAttributedString.Key.kern: 5.0 ],
            range: NSRange(location: 0, length: noAnnouncementText.count)
        )
        self.lbNoAnnouncement.attributedText = attributedNoAnnouncementText
        self.lbNoAnnouncement.textColor = Constants.appConfigColor.AnnouncementNoContentTextColor

        Constants.SendFib("AnnounceTableViewController")
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if let navController = self.navigationController {
            self.controllerTopStart = navController.navigationBar.frame.size.height
        }
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.refresh()
    }

    @objc func refresh() {
        self.loaded = false
        self.refreshControl.beginRefreshing()
        OPassAPI.GetAnnouncement(Constants.EventId) { (success: Bool, data: Any?, _) in
            if data != nil {
                if success {
                    self.loaded = true
                    if let data = data as? [AnnouncementInfo] {
                        self.announceJsonArray = data
                    }
                    self.announceTableView.reloadData()
                } else {
                    if data != nil {
                        self.performSegue(
                            withIdentifier: "ShowInvalidNetworkMsg",
                            sender: NSLocalizedString("Networking_Broken", comment: "")
                        )
                    } else {
                        self.loaded = true
                        self.announceJsonArray = [AnnouncementInfo]()
                        self.announceTableView.reloadData()
                    }
                }
            }
            self.refreshControl.endRefreshing()
        }
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let destination = segue.destination
        if destination.isMember(of: InvalidNetworkMessageViewController.self) {
            if let inmvc = destination as? InvalidNetworkMessageViewController {
                if let sender = sender as? String {
                    inmvc.message = sender
                    inmvc.delegate = self
                }
            }
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // MARK: - UITableViewControllerDelegate

    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let count = self.announceJsonArray.count
        if self.loaded {
            let NoAnnouncement = count == 0
            self.ivNoAnnouncement.isHidden = !NoAnnouncement
            self.lbNoAnnouncement.isHidden = !NoAnnouncement
        } else {
            self.ivNoAnnouncement.isHidden = true
            self.lbNoAnnouncement.isHidden = true
        }
        return count
    }

    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 12
    }

    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        view.tintColor = UIColor.clear
    }

    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 12
    }

    func tableView(_ tableView: UITableView, willDisplayFooterView view: UIView, forSection section: Int) {
        view.tintColor = UIColor.clear
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        guard let cell = tableView.dequeueReusableCell(withIdentifier: "AnnounceCell", for: indexPath) as? AnnounceTableViewCell else { return tableView.dequeueReusableCell(withIdentifier: "AnnounceCell", for: indexPath) }
        self.configureCell(cell, atIndexPath: indexPath)
        return cell
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return tableView.fd_heightForCell(withIdentifier: "AnnounceCell", configuration: { (cell: Any) in
            if let cell = cell as? AnnounceTableViewCell {
                self.configureCell(cell, atIndexPath: indexPath)
            }
        })
    }

    func configureCell(_ cell: AnnounceTableViewCell, atIndexPath indexPath: IndexPath) {
        cell.fd_enforceFrameLayout = false // Enable to use "-sizeThatFits:"
        cell.selectionStyle = .none
        cell.clipsToBounds = false
        cell.backgroundColor = UIColor.clear
        cell.layer.zPosition = CGFloat(indexPath.row)
        guard let vwContent = cell.vwContent else { return }
        vwContent.layer.cornerRadius = 5.0
        vwContent.layer.masksToBounds = true

        guard let vwShadowContent = cell.vwShadowContent else { return }
        vwShadowContent.layer.cornerRadius = 5.0
        vwShadowContent.layer.masksToBounds = false
        vwShadowContent.layer.shadowRadius = 50.0
        vwShadowContent.layer.shadowOffset = CGSize(width: 0, height: 50)
        vwShadowContent.layer.shadowColor = UIColor.black.cgColor
        vwShadowContent.layer.shadowOpacity = 0.1

        let announce = self.announceJsonArray[indexPath.row]
        guard let language = Bundle.main.preferredLocalizations.first else { return }

        cell.lbMessage.text = (language.contains("zh") ? announce.MsgZh : announce.MsgEn) + "\n"
        let uri = announce.URI
        let hasURL = uri != ""
        // cell.accessoryType = hasURL ? .disclosureIndicator : .none
        let datetime = announce.DateTime
        let strDate = Constants.DateToDisplayDateAndTimeString(datetime)
        cell.lbMessageTime.text = strDate
        cell.lbMessageTime.textColor = Constants.appConfigColor.AnnouncementSectionTitleTextColor
        cell.vwMessageTime.backgroundColor = Constants.appConfigColor.AnnouncementSectionTitleBackgroundColor

        if hasURL {
            cell.lbURL.text = uri
            if let foregroundColor = cell.lbIconOfURL.textColor {
            let titleAttribute: [NSAttributedString.Key: Any] = [
                    .font: Constants.fontOfAwesome(withSize: 20, inStyle: .solid),
                    .foregroundColor: foregroundColor,
                ]
                if let str = Constants.fontAwesome(code: "fa-external-link-alt") {
                    let title = NSAttributedString.init(
                        string: str,
                        attributes: titleAttribute
                    )
                    cell.lbIconOfURL.attributedText = title
                }
            }
        } else {
            cell.lbURL.text = ""
            cell.lbIconOfURL.attributedText = nil
        }
        cell.vwDashedLine.addDashedLine(Constants.appConfigColor.AnnouncementDashedLineColor)
        cell.vwURL.fd_collapsed = !hasURL
    }

    // MARK: - Table view delegate

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)

        let announce = self.announceJsonArray[indexPath.row]
        let uri = announce.URI

        if uri == "" {
            return
        }

        Constants.OpenInAppSafari(forPath: uri)

        Constants.SendFib("AnnounceTableView", WithEvents: ["URL": uri])
    }

    /*
     // Override to support conditional editing of the table view.
     - (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
     // Return NO if you do not want the specified item to be editable.
     return YES;
     }
     */

    /*
     // Override to support editing the table view.
     - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
     if (editingStyle == UITableViewCellEditingStyleDelete) {
     // Delete the row from the data source
     [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
     } else if (editingStyle == UITableViewCellEditingStyleInsert) {
     // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
     }
     }
     */

    /*
     // Override to support rearranging the table view.
     - (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
     }
     */

    /*
     // Override to support conditional rearranging of the table view.
     - (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
     // Return NO if you do not want the item to be re-orderable.
     return YES;
     }
     */

    /*
     #pragma mark - Navigation

     // In a storyboard-based application, you will often want to do a little preparation before navigation
     - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
     // Get the new view controller using [segue destinationViewController].
     // Pass the selected object to the new view controller.
     }
     */
}
